### PR TITLE
Use DI for controller communication rather than context

### PR DESCRIPTION
**Pull request by Gudahtt**
We now setup inter-controller communication using dependency injection rather than the `context` property. Any dependency a controller has is passed in via a constructor parameter. This was done in preparation for migrating to BaseControllerV2 and the new controller messaging system - it's just a temporary solution that will let us migrate controllers one at a time.

The style of dependency injection here matches the extension (at least with the newer controllers anyway). Specific methods and state snapshots are injected rather than entire controllers, to help simplify unit tests and make it easier to understand how controllers interact.

The mobile PR that demonstrates this is here: https://github.com/MetaMask/metamask-mobile/pull/2416

### CODE REVIEW

**Comment by Gudahtt** [comment]
TODO:
- [ ] Move additional constructor parameters into config or options bag
- [ ] Bring test coverage back up to 100%
- [ ] Reduce controller interactions
- [ ] Use stubs in tests rather than real controllers

File: src/assets/AssetsController.ts [hunk 1]
```diff
@@ -190,24 +189,6 @@ export class AssetsController extends BaseController<AssetsConfig, AssetsState>
...
   }
 
-  /**
-   * Get collectible tokenURI API following ERC721
-   *
-   * @param contractAddress - ERC721 asset contract address
-   * @param tokenId - ERC721 asset identifier
-   * @returns - Collectible tokenURI
-   */
-  private async getCollectibleTokenURI(contractAddress: string, tokenId: number): Promise<string> {
```
Comment (Gudahtt) [inline src/assets/AssetsController.ts:200]: This should be split into a separate PR. I found that this is the only place that `assetsContract.getCollectibleTokenURI` was called, so the prior check to ensure it supports metadata has been built into `assetsContract.getCollectibleTokenURI` to simplify this interaction.

File: src/assets/AssetsDetectionController.ts [hunk 2]
```diff
@@ -62,12 +61,12 @@ export class AssetsDetectionController extends BaseController<AssetsDetectionCon
   private async getOwnerCollectibles() {
     const { selectedAddress } = this.config;
     const api = this.getOwnerCollectiblesApi(selectedAddress);
-    const assetsController = this.context.AssetsController as AssetsController;
     let response: Response;
     try {
+      const openSeaApiKey = this.getOpenSeaApiKey();
```
Comment (Gudahtt) [inline src/assets/AssetsDetectionController.ts:66]: This was moved to a getter to make it easier to inject. This _could_ be done in a separate PR :thinking:

File: src/index.ts [hunk 3]
```diff
@@ -8,7 +8,6 @@ export * from './assets/AssetsContractController';
 export * from './assets/AssetsController';
 export * from './assets/AssetsDetectionController';
 export * from './BaseController';
-export * from './ComposableController';
```
Comment (Gudahtt) [inline src/index.ts:11]: This was a mistake, and should be undone.

File: src/user/PreferencesController.ts [hunk 4]
```diff
@@ -188,7 +190,25 @@ export class PreferencesController extends BaseController<BaseConfig, Preference
       };
       return ids;
     }, {});
-    this.update({ identities: { ...identities } });
+
+    let { selectedAddress } = this.state;
+    if (!Object.keys(identities).includes(selectedAddress)) {
+      selectedAddress = Object.keys(identities)[0];
+    }
+    this.update({ identities: { ...identities }, selectedAddress });
```
Comment (Gudahtt) [inline src/user/PreferencesController.ts:198]: The `updateIdentities` method has been updated to also reset the selected address, if the current selected address is no longer preset as an identity. This happens after every call to `updateIdentities`, so it may as well be built-in. This should be done in a separate PR.

File: src/user/PreferencesController.ts [hunk 5]
```diff
@@ -188,7 +190,25 @@ export class PreferencesController extends BaseController<BaseConfig, Preference
...
+  }
+
+  /**
+   * Update the selected address.
+   *
+   * @param address - The new selected address
+   */
+  updateSelectedAddress(address: string) {
```
Comment (Gudahtt) [inline src/user/PreferencesController.ts:206]: This was added so that other controllers could update the selected address without using `update` directly, which would allow them to set the state to something invalid. This should be added in a separate PR.

File: src/message-manager/TypedMessageManager.ts [hunk 6]
```diff
@@ -78,11 +78,6 @@ export class TypedMessageManager extends AbstractMessageManager<
    */
   name = 'TypedMessageManager';
 
-  /**
-   * List of required sibling controllers this controller needs to function
-   */
-  requiredControllers = ['NetworkController'];
```
Comment (Gudahtt) [inline src/message-manager/TypedMessageManager.ts:84]: This controller doesn't use the network controller. This should be removed in a separate PR.

File: src/assets/AssetsController.ts (hunk 1 above)
Comment (Gudahtt) [reply src/assets/AssetsController.ts:200]: Done in #414

File: src/assets/AssetsDetectionController.ts (hunk 2 above)
Comment (Gudahtt) [reply src/assets/AssetsDetectionController.ts:66]: Nah, I'll leave it. It seems like a bit of a silly change in isolation.

File: src/user/PreferencesController.ts (hunk 4 above)
Comment (Gudahtt) [reply src/user/PreferencesController.ts:198]: Done in #415

File: src/user/PreferencesController.ts (hunk 5 above)
Comment (Gudahtt) [reply src/user/PreferencesController.ts:206]: Oh, this method already exists :facepalm: It's below, `setSelectedAddress`. I'll remove this.

File: src/message-manager/TypedMessageManager.ts (hunk 6 above)
Comment (Gudahtt) [reply src/message-manager/TypedMessageManager.ts:84]: Removed in #416

File: src/assets/AccountTrackerController.ts [hunk 7]
```diff
@@ -71,24 +68,40 @@ export class AccountTrackerController extends BaseController<AccountTrackerConfi
    */
   name = 'AccountTrackerController';
 
-  /**
-   * List of required sibling controllers this controller needs to function
-   */
-  requiredControllers = ['PreferencesController'];
+  private identities: PreferencesState['identities'];
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/assets/AccountTrackerController.ts:71]: It wasn't intended to be removed :thinking: I will investigate

File: src/assets/AssetsDetectionController.ts [hunk 8]
```diff
@@ -104,6 +144,25 @@ export class AssetsDetectionController extends BaseController<AssetsDetectionCon
       tokens: [],
     };
     this.initialize();
+    this.getAssetsState = getAssetsState;
+    this.addTokens = addTokens;
+    onAssetsStateChange(({ tokens }) => {
+      this.configure({ tokens });
+    });
+    onPreferencesStateChange(({ selectedAddress }) => {
+      const actualSelectedAddress = this.config.selectedAddress;
```
Comment (estebanmino) [inline src/assets/AssetsDetectionController.ts:153]: just a question here, config is going to be deprecated in next updates?

Comment (Gudahtt) [reply src/assets/AssetsDetectionController.ts:153]: It will be removed in the new base controller. But we'll be migrating each controller to that one at a time. Any controllers still extending the old base controller will continue to use config in the interim.

Though, in this particular case, I'm not sure we need to use the config to store `selectedAddress`  :thinking:

**Comment by Gudahtt** [comment]
I've just rebased this and resolved the test conflicts - I haven't addressed the comments yet. Looking at those now.

File: src/ComposableController.ts [hunk 9]
```diff
@@ -36,45 +22,22 @@ export class ComposableController extends BaseController<any, any> {
...
     this.controllers = controllers;
-    this.cachedState = undefined;
-  }
-
-  /**
-   * Get current list of child composed store instances
-   *
-   * @returns - List of names to controller instances
-   */
-  get controllers() {
```
Comment (Gudahtt) [inline src/ComposableController.ts:52]: Note that the getter and setter for the `controllers` property has been removed. It appears to have been completely unused. I searched and found no references to it in this repo or in mobile.

File: src/assets/AccountTrackerController.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/assets/AccountTrackerController.ts:71]: I couldn't find any relevant changes here to explain what you're describing, but I suspect I found the cause of this problem in the mobile PR: https://github.com/MetaMask/metamask-mobile/pull/2416/files#r599841257

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/assets/AccountTrackerController.ts:71]: I'm going to update this anyway to replace it with `getPreferencesState` function. That might be a bit easier to understand, and more in-line with how the future controller messaging system works.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/assets/AccountTrackerController.ts]: Resolved here: a3b6d00132497f9fbfc1aa0f93573ab8aa021ce7

File: src/assets/AssetsDetectionController.ts (hunk 8 above)
Comment (Gudahtt) [reply src/assets/AssetsDetectionController.ts:153]: I'm going to leave this as-is for now, as keeping this in the config does still serve a purpose. It makes it easier to see when the address has changed, v.s. other state. This is easier to detect with the new base controller API, but with the current one this is a reasonable solution.

**Comment by Gudahtt** [comment]
This has just been rebased to fix some conflicts with #426 and #435

File: src/ComposableController.ts [hunk 10]
```diff
@@ -36,45 +22,22 @@ export class ComposableController extends BaseController<any, any> {
...
-   */
-  set controllers(controllers: ControllerList) {
-    this.internalControllers = controllers;
-    const initialState: any = {};
-    controllers.forEach((controller) => {
+    this.controllers.forEach((controller) => {
       const { name } = controller;
-      this.context[name] = controller;
-      controller.context = this.context;
-      this.cachedState?.[name] && controller.update(this.cachedState[name]);
```
Comment (Gudahtt) [inline src/ComposableController.ts:68]: Apparently this line is how most controllers on mobile had their initial state set. I had wrongly assumed they were being passed initial state via their constructors.

This initial state setting has been re-implemented in mobile: https://github.com/MetaMask/metamask-mobile/pull/2416/commits/2b8023568161ad9d31b0e525f6c3318e71a436e5
Going forward we should set initial state in the constructor instead.

File: src/assets/AssetsDetectionController.ts [hunk 11]
```diff
@@ -84,18 +83,59 @@ export class AssetsDetectionController extends BaseController<AssetsDetectionCon
...
+   * @param options
+   * @param options.onAssetsStateChange - Allows subscribing to assets controller state changes
+   * @param options.onPreferencesStateChange - Allows subscribing to preferences controller state changes
+   * @param options.onNetworkStateChange - Allows subscribing to network controller state changes
+   * @param options.getOpenSeaApiKey - Gets the OpenSea API key, if one is set
+   * @param options.getBalancesInSingleCall - Gets the balances of a list of tokens for the given address
+   * @param options.addTokens - Add a list of tokens
+   * @param options.addCollectible - Add a collectible
+   * @param options.removeCollectible - Remove a collectible
+   * @param options.initialAssetsState - The initial state of the Assets controller
```
Comment (Gudahtt) [inline src/assets/AssetsDetectionController.ts:110]: This should be updated - `initialAssetsState` was replaced with `getAssetsState`

**Review by shanejonas — DISMISSED** [review]

File: src/assets/AssetsDetectionController.ts (hunk 11 above)
Comment (Gudahtt) [reply src/assets/AssetsDetectionController.ts]: Resolved in c991da5f5e24341742af5a36efa57053459e410a

File: src/user/PreferencesController.ts [hunk 12]
```diff
@@ -47,7 +47,6 @@ export interface PreferencesState extends BaseState {
   lostIdentities: { [address: string]: ContactEntry };
   selectedAddress: string;
 }
-
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/user/PreferencesController.ts]: Resolved in c991da5f5e24341742af5a36efa57053459e410a

**Review by shanejonas — APPROVED** [review]

**Review by estebanmino — APPROVED** [review]

**Comment by Gudahtt** [comment]
This has been rebased to resolve a conflict with #439

**Review by estebanmino — APPROVED** [review]
👌
